pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -154,24 +163,91 @@ func (p *Pod) isUnretriableGroup() bool {
 
 // IsSuspended returns whether the job is suspended or not.
 func (p *Pod) IsSuspended() bool {
-	return podSuspended(&p.pod)
+	if !p.isGroup {
+		return podSuspended(&p.pod)
+	}
+
+	for i := range p.list.Items {
+		if podSuspended(&p.list.Items[i]) {
+			return true
+		}
+	}
+
+	return false
 }
 
 // Suspend will suspend the job.
 func (p *Pod) Suspend() {
 	// Not implemented because this is not called when JobWithCustomStop is implemented.
 }
 
-// RunWithPodSetsInfo will inject the node affinity and podSet counts extracting from workload to job and unsuspend it.
-func (p *Pod) RunWithPodSetsInfo(podSetsInfo []podset.PodSetInfo) error {
-	if p.groupName() == "" && len(podSetsInfo) != 1 {
-		return fmt.Errorf("%w: expecting 1 pod set got %d", podset.ErrInvalidPodsetInfo, len(podSetsInfo))
-	}
-	idx := gateIndex(&p.pod)
+// ungatePod removes the kueue scheduling gate from the pod.
+// Returns true if the pod has been ungated and false otherwise.
+func ungatePod(pod *corev1.Pod) bool {
+	idx := gateIndex(pod)
 	if idx != gateNotFound {
-		p.pod.Spec.SchedulingGates = append(p.pod.Spec.SchedulingGates[:idx], p.pod.Spec.SchedulingGates[idx+1:]...)
+		pod.Spec.SchedulingGates = append(pod.Spec.SchedulingGates[:idx], pod.Spec.SchedulingGates[idx+1:]...)
+		return true
+	}
+
+	return false
+}
+
+// Run will inject the node affinity and podSet counts extracting from workload to job and unsuspend it.
+func (p *Pod) Run(ctx context.Context, c client.Client, podSetsInfo []podset.PodSetInfo) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	if !p.isGroup {
+		if len(podSetsInfo) != 1 {
+			return fmt.Errorf("%w: expecting 1 pod set got %d", podset.ErrInvalidPodsetInfo, len(podSetsInfo))
+		}
+
+		if ungated := ungatePod(&p.pod); !ungated {
+			return nil
+		}
+
+		if err := podset.Merge(&p.pod.ObjectMeta, &p.pod.Spec, podSetsInfo[0]); err != nil {
+			return err
+		}
+
+		return c.Update(ctx, &p.pod)
+	}
+
+	for _, podInGroup := range p.list.Items {
+		if ungated := ungatePod(&podInGroup); !ungated {
+			continue
+		}
+
+		roleHash, err := getRoleHash(podInGroup)
+		if err != nil {
+			return err
+		}
+
+		podSetIndex := slices.IndexFunc(podSetsInfo, func(info podset.PodSetInfo) bool {
+			return info.Name == roleHash
+		})
+		if podSetIndex == -1 {
```

Comment:
Shouldn't we delete this violating Pod? Or is it already deleted somewhere else?

Reply:
It should be deleted by `FindMatchingWorkloads` on the next reconcile.